lib/syskit/exceptions.rb

```diff
@@ -447,37 +449,49 @@ def initialize(device, task0, task1, toplevel_tasks_to_requirements = {})
             @tasks = [task0, task1]
 
-            solver = NetworkGeneration::MergeSolver.new(task0.plan)
-            @merge_result = solver.resolve_merge(task0, task1, {})
             @involved_definitions = @tasks.map do |t|
                 find_all_related_syskit_actions(t, toplevel_tasks_to_requirements)
             end
         end
 
-        def find_all_related_syskit_actions(task, toplevel_tasks_to_requirements)
-            result = []
-            while task
-                result.concat(toplevel_tasks_to_requirements[task] || [])
-                task = task.each_parent_task.first
-            end
-            result
-        end
-
         def pretty_print(pp)
             pp.text "device '#{device.name}' of type #{device.model} is assigned "
             pp.text "to two tasks that cannot be merged"
-            pp.breakable
-            @merge_result.pretty_print_failure(pp)
-            @involved_definitions.each_with_index do |defs, i|
-                next if defs.empty?
+            print_failed_merge_chain(pp, *@tasks)
+            @tasks.zip(@involved_definitions).each do |t, defs|
+                print_dependent_definitions(pp, t, defs)
+            end
+        end
+    end
 
-                pp.breakable
-                pp.text "Chain #{i + 1} is needed by the following definitions:"
-                pp.nest(2) do
-                    defs.each do |d|
-                        pp.breakable
-                        pp.text d.to_s
-                    end
+    class ConflictingDeploymentAllocation < SpecError
+        include Syskit::NetworkGenerationsExceptionHelpers
+
+        attr_reader :deployment_to_tasks, :toplevel_tasks_to_requirements
+
+        def initialize(deployment_to_tasks, toplevel_tasks_to_requirements = {})
+            @deployment_to_tasks = deployment_to_tasks
+            @toplevel_tasks_to_requirements = toplevel_tasks_to_requirements
+        end
+
+        def pretty_print(pp)
+            deployment_to_tasks.each do |deployed_task, tasks|
+                process_server_name = deployed_task.configured_deployment
+                                                   .process_server_name
+                orogen_model = deployed_task.configured_deployment.orogen_model
+                pp.text(
+                    "deployed task '#{deployed_task.mapped_task_name}' from deployment " \
+                    "'#{orogen_model.name}' defined in " \
+                    "'#{orogen_model.project.name}' on '#{process_server_name}' is " \
+                    "assigned to multiple tasks. Here follows one merge failure " \
+                    "(it can have more):"
+                )
+                print_failed_merge_chain(pp, tasks[0], tasks[1])
+                tasks.each do |t|
+                    defs = find_all_related_syskit_actions(
+                        t, toplevel_tasks_to_requirements
+                    )
+                    print_dependent_definitions(pp, t, defs)
                 end
```

Comment:
I think it would make more sense to display the list of tasks first, and then the merge failure:

```
assigned to #{tasks.size} tasks. After the list is a detailed explanation of why the first two tasks are not merged:
```